services/api/lib/postgre.js

```diff
@@ -295,6 +295,334 @@ module.exports = function (pg) {
+              case 'create': {
+                values = [
+                  data.projectId,
+                  userId,
+                  data.x,
+                  data.y,
+                  JSON.stringify(data.styles || {})
+                ];
+                break;
+              }
+              case 'update': {
+                values = [
+                  data.x,
+                  data.y,
+                  JSON.stringify(data.styles || {}),
+                  data.id
+                ];
+                break;
+              }
+              case 'remove': {
+                values = [
+                  data.id
+                ];
+                break;
+              }
+            }
+            break;
+          }
+          case 'elements': {
+            switch (method) {
+              case 'create': {
+                values = [
+                  data.pageId,
+                  userId,
+                  data.type,
+                  JSON.stringify(data.attributes || {}),
+                  JSON.stringify(data.styles || {})
+                ];
+                break;
+              }
+              case 'update': {
+                values = [
+                  JSON.stringify(data.styles || {}),
+                  JSON.stringify(data.attributes || {}),
+                  data.id
+                ];
+                break;
+              }
+              case 'remove': {
+                values = [
+                  data.id
+                ];
+                break;
+              }
+            }
+          }
+        }
+        return values;
+      }
+
+      function hasPermissionsForAction(userId, type, method, data, results, transaction) {
+        return new Promise(function(resolve, reject) {
+          var lookupType;
+          var lookupId;
+
+          // determine the type of query to run, and the id value to look up
+          if ( method === 'create' ) {
+            switch (type) {
+              case 'projects': {
+                // the user should always be able to create new projects under their account, so resolve true now
+                return resolve(true);
+              }
+              case 'pages': {
+                // look up the project that this page should be created for, to ensure the current user owns it
+                lookupType = 'projects';
+                lookupId = data.projectId;
+                break;
+              }
+              case 'elements': {
+                // look up the page this element is to be created for, to ensure the current user owns it
+                lookupType = 'pages';
+                lookupId = data.pageId;
+                break;
+              }
+            }
+          } else {
+            // update / delete need to look up the target type by id to verify ownership
+            lookupType = type;
+            lookupId = data.id;
+          }
+
+          // coerce these values to string, because bigints come in from PG as strings
```

Comment:
From what I've read on the interwebz, both methods of turning a number to a string are similar in cost. toString may be a little friendlier to non-JS programmers, but it's how numbers are coerced is basically the entire app, so for now I'll stick to that method, and if you reaaaaally have a project wide nit, we can update it all in a new ticket